src/dmd/semantic2.d

```diff
@@ -596,3 +614,101 @@ private extern(C++) final class Semantic2Visitor : Visitor
+ *            only one ABI tag object is allowed per symbol
+ *            (but it can have multiple tags as it's an array exp).
+ */
+private void doGNUABITagSemantic(ref Expression e, ref Expression* lastTag)
+{
+    import dmd.dmangle;
+
+    // When `@gnuAbiTag` is used, the type will be the UDA, not the struct literal
+    if (e.op == TOK.type)
+    {
+        e.error("`@%s` at least one argument expected", Id.udaGNUAbiTag.toChars());
+        return;
+    }
+
+    // Definition is in `core.attributes`. If it's not a struct literal,
+    // it shouldn't have passed semantic, hence the `assert`.
+    auto sle = e.isStructLiteralExp();
+    if (sle is null)
+    {
+        assert(global.errors);
+        return;
+    }
+    // The definition of `gnuAttributes` only have 1 member, `string[] tags`
+    assert(sle.elements && sle.elements.length == 1);
+    // `gnuAbiTag`'s constructor is defined as `this(string[] tags...)`
+    auto ale = (*sle.elements)[0].isArrayLiteralExp();
+    if (ale is null)
+    {
+        e.error("`@%s` at least one argument expected", Id.udaGNUAbiTag.toChars());
+        return;
+    }
+
+    // Check that it's the only tag on the symbol
+    if (lastTag !is null)
+    {
+        const str1 = (*lastTag.isStructLiteralExp().elements)[0].toString();
+        const str2 = ale.toString();
+        e.error("only one `@%s` allowed per symbol", Id.udaGNUAbiTag.toChars());
+        e.errorSupplemental("instead of `@%s @%s`, use `@%s(%.*s, %.*s)`",
+            lastTag.toChars(), e.toChars(), Id.udaGNUAbiTag.toChars(),
+            // Avoid [ ... ]
+            cast(int)str1.length - 2, str1.ptr + 1,
+            cast(int)str2.length - 2, str2.ptr + 1);
+        return;
+    }
+    lastTag = &e;
+
+    // We already know we have a valid array literal of strings.
+    // Now checks that elements are valid.
+    foreach (idx, elem; *ale.elements)
+    {
+        const str = elem.toStringExp().peekString();
+        if (!str.length)
+        {
+            e.error("argument `%d` to `@%s` cannot be %s", cast(int)(idx + 1),
+                    Id.udaGNUAbiTag.toChars(),
+                    elem.isNullExp() ? "`null`".ptr : "empty".ptr);
+            continue;
+        }
+
+        foreach (c; str)
+        {
+            if (!c.isValidMangling())
+            {
+                e.error("`@%s` char `0x%02x` not allowed in mangling",
+                        Id.udaGNUAbiTag.toChars(), c);
+                break;
+            }
+        }
+        // Valid element
+    }
+    // Since ABI tags need to be sorted, we sort them in place
+    // It might be surprising for users that inspects the UDAs,
+    // but it's a concession to practicality.
+    // Casts are unfortunately necessary as `implicitConvTo` is not
+    // `const` (and nor is `StringExp`, by extension).
+    static int predicate(const scope Expression* e1, const scope Expression* e2) nothrow
+    {
+        scope(failure) assert(0, "An exception was thrown");
```

Comment:
dmd doesn't throw, so why this?

Reply:
Might be that some of those functions are not annotated as `nothrow`, will double check and fix as needed.

Reply:
Okay I realize why I didn't do this: It's a giant rabbit hole. Same issue with `const`. You'd have to make most of DMD `nothrow` for it to work, and I just felt it was too much of a change.
Also, consider that in order for this to work, the visitors need to be `nothrow` / whatever attributes we fancy, which greatly reduce the usability of DMD as a library.

Reply:
Note that this should *never* be triggered. Indeed DMD does not `throw`. `scope (failure) assert(0);` is a cheap trick to make the function `nothrow`.